File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
# TC- O(logk) where k is max(nums)
# SC- O(1)

class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        
        def caneat(k):
            hours = 0
            for pile in piles:
                hours += math.ceil(pile/k)
                if hours>h:
                    return False
            return True

        # binary search in ans space
        left, right = 1, max(piles)
        while left<right:
            mid = left+(right-left)//2

            if(caneat(mid)):
                right = mid
            else:
                left = mid+1
        return left

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.cpp
// TC- O(nlog(max(piles)))
// SC- O(1)

class Solution {
public:
    bool canEat(const int speed, const int h, const vector<int>& piles){
        int total_h = 0;
        for(int pile:piles){
            total_h += pile/speed;
            if(pile%speed!=0)
                total_h ++;
        }
        return total_h<=h;
    }

    int minEatingSpeed(vector<int>& piles, int h) {
        int left = 1, right = 0;
        for(int pile:piles)
            right = max(right, pile);
        
        while(left<right){
            int mid = left+(right-left)/2;
            if(canEat(mid, h, piles))
                right = mid;
            else
                left = mid+1;
        }
        return left;
    }
};